authority/admin/api/webhook.go

```diff
@@ -0,0 +1,224 @@
+		return
+	}
+	if newWebhook.Secret != "" {
+		err := admin.NewError(admin.ErrorBadRequestType, "webhook secret must not be set")
+		render.Error(w, err)
+		return
+	}
+
+	id, err := randutil.UUIDv4()
+	if err != nil {
+		render.Error(w, admin.WrapErrorISE(err, "error generating webhook id"))
+		return
+	}
+	newWebhook.Id = id
+
+	// verify the name is unique
+	for _, wh := range prov.Webhooks {
+		if wh.Name == newWebhook.Name {
+			err := admin.NewError(admin.ErrorConflictType, "provisioner %s already has a webhook with the name %s", prov.Name, newWebhook.Name)
+			render.Error(w, err)
+			return
+		}
+	}
+
+	secret, err := randutil.Bytes(64)
+	if err != nil {
+		render.Error(w, admin.WrapErrorISE(err, "error generating webhook secret"))
+		return
+	}
+	newWebhook.Secret = base64.StdEncoding.EncodeToString(secret)
+
+	prov.Webhooks = append(prov.Webhooks, newWebhook)
+
+	auth := mustAuthority(ctx)
+	if err := auth.UpdateProvisioner(ctx, prov); err != nil {
+		if isBadRequest(err) {
+			render.Error(w, admin.WrapError(admin.ErrorBadRequestType, err, "error creating provisioner webhook"))
+			return
+		}
+
+		render.Error(w, admin.WrapErrorISE(err, "error creating provisioner webhook"))
+		return
+	}
+
+	render.ProtoJSONStatus(w, newWebhook, http.StatusCreated)
+}
+
+func (war *webhookAdminResponder) DeleteProvisionerWebhook(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	prov := linkedca.MustProvisionerFromContext(ctx)
+
+	webhookName := chi.URLParam(r, "webhookName")
+
+	found := false
+	for i, wh := range prov.Webhooks {
+		if wh.Name == webhookName {
+			prov.Webhooks = append(prov.Webhooks[0:i], prov.Webhooks[i+1:]...)
+			found = true
+			break
+		}
+	}
+	if !found {
+		render.JSONStatus(w, DeleteResponse{Status: "ok"}, http.StatusOK)
+		return
+	}
+
+	auth := mustAuthority(ctx)
+	if err := auth.UpdateProvisioner(ctx, prov); err != nil {
+		if isBadRequest(err) {
+			render.Error(w, admin.WrapError(admin.ErrorBadRequestType, err, "error deleting provisioner webhook"))
+			return
+		}
+
+		render.Error(w, admin.WrapErrorISE(err, "error deleting provisioner webhook"))
+		return
+	}
+
+	render.JSONStatus(w, DeleteResponse{Status: "ok"}, http.StatusOK)
+}
+
+func (war *webhookAdminResponder) UpdateProvisionerWebhook(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	prov := linkedca.MustProvisionerFromContext(ctx)
+
+	var newWebhook = new(linkedca.Webhook)
+	if err := read.ProtoJSON(r.Body, newWebhook); err != nil {
+		render.Error(w, err)
+		return
+	}
+
+	if err := validateWebhook(newWebhook); err != nil {
+		render.Error(w, err)
+		return
+	}
+
+	found := false
+	for i, wh := range prov.Webhooks {
+		if wh.Name != newWebhook.Name {
+			continue
+		}
+		if newWebhook.Secret != "" && newWebhook.Secret != wh.Secret {
+			err := admin.NewError(admin.ErrorBadRequestType, "webhook secret cannot be updated")
+			render.Error(w, err)
+			return
+		}
+		newWebhook.Secret = wh.Secret
+		newWebhook.Id = wh.Id
```

Comment:
a938904
I enforced blank ID or secret. I don't want to add that to validate because I want clients to be able to use the response from create in an update request without having to delete the secret and id.